CopyDataObjectFilter

jmarquisbq opened the pull request:
- CopyDataGroupFilter refactored and updated to be able to copy any DataObject instead of just groups
- Filter can now copy multiple DataObjects at once instead of just a single object
    - Added a MultiPathSelectionParameter for selecting multiple DataObjects (of any type) at once 



<!-- The text within this markup is a comment, and is intended to provide
guidelines to open a Pull Request for the complex repository. This text will not
be part of the Pull Request. -->


<!-- See the CONTRIBUTING (CONTRIBUTING.md) guide. Specifically:

Start complex commit messages with a standard prefix (and a space):

 * BUG: fix for runtime crash or incorrect result
 * COMP: compiler error or warning fix
 * DOC: documentation change
 * ENH: new functionality
 * PERF: performance improvement
 * STYLE: no logic impact (indentation, comments)
 * WIP: Work In Progress not ready for merge

Provide a short, meaningful message that describes the change you made.

When the PR is based on a single commit, the commit message is usually left as
the PR message.

A reference to a related issue or pull request (https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)
in your repository. You can automatically
close a related issues using keywords (https://help.github.com/articles/closing-issues-using-keywords/)

@mentions (https://help.github.com/articles/basic-writing-and-formatting-syntax/#mentioning-people-and-teams)
of the person or team responsible for reviewing proposed changes. -->


## Naming Conventions

Naming of variables should descriptive where needed. Loop Control Variables can use `i` if warranted. Most of these conventions are enforced through the clang-tidy and clang-format configuration files.

- [x] Class and Structs are `UpperCamelCase`
- [x] Class private member variables are `m_UpperCamelCase`
- [x] Class methods are `lowerCamelCase`
- [x] Method arguments are `lowerCamelCase`
- [x] Normal variables are `lowerCamelCase`
- [x] Constants are `k_UpperCamelCase`
- [x] Global statics are `s_UpperCamelCase`
- [x] Free Functions are `UpperCamelCase`
- [x] Macros are `ALL_UPPER_SNAKE_CASE`
- [x] Unit test will test data output from the filter.
- [x] Reused strings should be constants in an anonymous namespace
- [x] If parallelization is used, proper use of the abstracted complex classes are used. Using TBB specifically in a filter should be frowned upon unless for a really good reason.

## Filter Checklist

- [x] Parameters should be generally broken down into "Input Parameters", "Required Data Objects", "Created Data Objects". There can be exceptions to this.
- [x] ChoicesParameter selections should be an enumeration defined in the filer header
- [x] Documentation copied from SIMPL Repo and updated (if necessary)
- [x] Parameter argument variables are k_CamelCase_Key
- [x] Parameter argument strings are lower_snake_case
```
static inline constexpr StringLiteral k_AlignmentType_Key = "alignment_type";
```

## Unit Testing
- [x] 1 Unit test to instantiate the filter with the default arguments.
- [x] 1 Unit test to test output from the filter against known exemplar set of data
- [x] 1 Unit test to test invalid input code paths that are specific to a filter. Don't test that a DataPath does not exist since that test is already performed as part of the SelectDataArrayAction.

## Code Cleanup
- [x] No commented out code (rare exceptions to this is allowed..)
- [x] Filters should have both the Filter class and Algorithm class for anything beyond trivial needs
- [x] No API changes were made (or the changes have been approved)
- [x] No major design changes were made (or the changes have been approved)
- [x] Added test (or behavior not changed)
- [x] Updated API documentation (or API not changed)
- [x] Added license to new files (if any)
- [x] Added Python wrapping to new files (if any) as necessary
- [x] Added example pipelines that use the filter
- [x] Classes and methods are properly documented.


<!-- **Thanks for contributing to complex!** -->

jmarquisbq committed:
Refactor & abstract CopyDataGroup to CopyDataObjectFilter

jmarquisbq committed:
Add MultiPathSelectionParameter for selecting multiple objects of any type

jmarquisbq committed:
Fix ArraySelectionParameter clone function

jmarquisbq committed:
Update CopyDataObject to allow copying multiple objects at once

jmarquisbq committed:
Check for empty suffix string when same parent group used

jmarquisbq committed:
Update the unit test

nyoungbq commented on src/Plugins/ComplexCore/src/ComplexCore/Filters/CopyDataObjectFilter.cpp:
```diff
@@ -0,0 +1,109 @@
+
+namespace complex
+{
+namespace
+{
+} // namespace
```
```suggestion
```
This empty anonymous namespace can be removed.

nyoungbq committed:
Update ComplexCoreLegacyUUIDMapping.hpp

Link the SIMPL UUID for CopyObject to CopyDataObjectFilter UUID in legacy UUID mapping

nyoungbq commented on src/Plugins/ComplexCore/test/CopyDataObjectTest.cpp:
```diff
@@ -0,0 +1,103 @@
+
+using namespace complex;
+
+TEST_CASE("ComplexCore::CopyDataObjectFilter(Instantiate)", "[ComplexCore][CopyDataObjectFilter]")
```
```suggestion
TEST_CASE("ComplexCore::CopyDataObjectFilter(Valid Execution)", "[ComplexCore][CopyDataObjectFilter]")
```

nyoungbq commented on src/complex/Parameters/MultiPathSelectionParameter.hpp:
```diff
@@ -0,0 +1,101 @@
+namespace complex
+{
+/**
+ * @brief This Filter Parameter represents multiple DataArrays as input.
```
```suggestion
 * @brief This Filter Parameter represents multiple DataObjects as input.
```

nyoungbq commented on src/complex/Parameters/MultiPathSelectionParameter.hpp:
```diff
@@ -0,0 +1,101 @@
+  std::any defaultValue() const override;
+
+  /**
+   * @brief
```
```suggestion
   * @brief Returns the default path.
```

nyoungbq commented on src/complex/Parameters/MultiPathSelectionParameter.hpp:
```diff
@@ -0,0 +1,101 @@
+  Result<> validate(const DataStructure& dataStructure, const std::any& value) const override;
+
+  /**
+   * @brief Validates the given value against the given DataStructure. Returns warnings/errors.
```
```suggestion
   * @brief Validates the given paths against the given DataStructure. Returns warnings/errors.
```

nyoungbq commented on src/complex/Parameters/MultiPathSelectionParameter.hpp:
```diff
@@ -0,0 +1,101 @@
+  /**
+   * @brief Takes the value and a mutable DataStructure and attempts store the actual derived DataObject in the std::any.
+   * Returns any warnings/errors.
+   * @param dataStructure
+   * @param value
```
```suggestion
   * @param dataStructure The active DataStructure to use during resolution
   * @param value The value to resolve
```

nyoungbq commented on src/complex/Parameters/MultiPathSelectionParameter.cpp:
```diff
@@ -0,0 +1,136 @@
+  std::vector<Error> errors;
+  for(usize i = 0; i < value.size(); i++)
+  {
+    const auto& path = value.at(i);
```
Did you mean to do a bounds check with the `.at(i)` ?
- If not switch to the bracket `[i]` operator for speed optimization.
- If you did I would add a comment to mention the bounds check is purposeful

jmarquisbq replied:
```suggestion
    const auto& path = value[i];
```

nyoungbq commented on src/complex/Parameters/MultiPathSelectionParameter.cpp:
```diff
@@ -0,0 +1,136 @@
+  nlohmann::json json = nlohmann::json::array();
+  for(const auto& path : paths)
+  {
+    json.push_back(path.toString());
```
```suggestion
    json.push_back(std::move(path.toString()));
```
I assume the string being returned is being assembled in the function and returned as an rvalue thus std::move ensures its an in-place move-construction preventing loss of efficiency

nyoungbq reviewed (commented):
Might want to make sure the documentation changes I offered up are correct. Also I am not sure if it was already done, because I don't see it in the files changed, but make sure the documentation for this filter is updated/brought in

jmarquisbq committed:
Remove unused namespace

Co-authored-by: Nathan Young <109472155+nyoungbq@users.noreply.github.com>

jmarquisbq committed:
Fix unit test name

Co-authored-by: Nathan Young <109472155+nyoungbq@users.noreply.github.com>

jmarquisbq committed:
Fix MultiPathSelectionParameter description comment

Co-authored-by: Nathan Young <109472155+nyoungbq@users.noreply.github.com>

jmarquisbq committed:
Fix MultiPathSelectionParameter function description comments

Co-authored-by: Nathan Young <109472155+nyoungbq@users.noreply.github.com>

jmarquisbq committed:
Apply suggestions from code review

Co-authored-by: Nathan Young <109472155+nyoungbq@users.noreply.github.com>

jmarquisbq committed:
Update CopyDataObjectFilter doc

imikejackson committed:
Merge branch 'develop' into filter/CopyDataObject
